onnxruntime/core/optimizer/bias_softmax_fusion.cc

```diff
@@ -0,0 +1,226 @@
+    *mask = add_node.MutableInputDefs()[1];
+  }
+  else {
+    *input = add_node.MutableInputDefs()[1];
+    *mask = add_node.MutableInputDefs()[0];
+  }
+}
+
+Status BiasSoftmaxFusion::ApplyImpl(Graph& graph, bool& modified, int graph_level, const logging::Logger& logger) const {
+  GraphViewer graph_viewer(graph);
+  const auto& node_topology_list = graph_viewer.GetNodesInTopologicalOrder();
+
+  // only support CUDA execution provider
+  auto& cep = GetCompatibleExecutionProviders();
+  if (cep.size() > 0 && cep.find(kCudaExecutionProvider) == cep.end())
+    return Status::OK();
+
+  for (auto node_index : node_topology_list) {
+    auto* node_ptr = graph.GetNode(node_index);
+    if (nullptr == node_ptr)
+      continue;  // node was removed
+
+    auto& node = *node_ptr;
+
+    ORT_RETURN_IF_ERROR(Recurse(node, modified, graph_level, logger));
+
+    // pattern match on dropout(softmax(input + mask)) subgraph
+    // -----------------------------------------------------------------------
+
+    // check node is add and has single output
+    if (!graph_utils::IsSupportedOptypeVersionAndDomain(node, "Add", {7}) ||
+        !graph_utils::IsSupportedProvider(node, {kCudaExecutionProvider}) ||
+        !optimizer_utils::CheckOutputEdges(graph, node, 1)) {
+      continue;
+    }
+
+    // check shape information is not available for both add inputs
+    auto& add_node = node;
+    NodeArg* input1 = add_node.MutableInputDefs()[0];
+    NodeArg* input2 = add_node.MutableInputDefs()[1];
+    const TensorShapeProto* S1 = input1->Shape();
+    const TensorShapeProto* S2 = input2->Shape();
+    if (S1 == nullptr || S2 == nullptr || S1->dim_size() < 1 || S2->dim_size() < 1) {
+      continue;
+    }
+
+    // check add is only consumed by softmax with matching exec provider
+    auto p = add_node.OutputNodesBegin();
+    if (p == add_node.OutputNodesEnd()) {
+      continue;
+    }
+    Node& softmax_node = const_cast<Node&>(*p);
+    if (!graph_utils::IsSupportedOptypeVersionAndDomain(softmax_node, "Softmax", {1, 11}) ||
+        softmax_node.GetExecutionProviderType() != add_node.GetExecutionProviderType()) {
+      continue;
+    }
+
+    // can't perform conversion if output is graph output
+    if (!graph.GetNodeOutputsInGraphOutputs(add_node).empty()) {
+      continue;
+    }
+
+    // check mask can broadcast across input batches with simple division
+    // -----------------------------------------------------------------------
+
+    /* Here we check input and mask dimensions are as expected:
+
+       Let 
+           input shape = [ a0 ... a(B-1) aB a(B+1) ... a(k-1) ak a(k+1) ... a(N-1) ] 
+       with rank = N and softmax axis = k and to be determined broadcast axis = B.
+
+       Then
+           mask shape  = [ a0 ... a(B-1) 1    1    ...   1    ak a(k+1) ... a(N-1) ] 
+       with rank = N and B < k, OR
+           mask shape  = [                   ...    1    1    ak a(k+1) ... a(N-1) ] 
+       with rank = N-k.
```

Comment:
if i'm understanding correctly, `rank = N-k + <number of 1's>`, right?
also `mask shape  = [                   ...    1    1    ak a(k+1) ... a(N-1) ]` might benefit from something to the left of the left-most `...`
this is a helpful bit of documentation though! 👍